### PR TITLE
WIP: add CuRowSparseMatrix class

### PR DESCRIPTION
**Pull request by naxingyu**
This PR is for #298

### CODE REVIEW

File: src/cudamatrix/cu-kernels.cu [hunk 1]
```diff
@@ -748,6 +748,27 @@ static void _add_mat_mat_elements(Real *data, const Real *srcA_data, const Real
...
+static void _add_mat_smat(Real *data, int stride, Real alpha, const Real *srcA_data, const int srcA_stride, bool transA, const RowElement<Real>* srcB_data, const int srcB_stride, Real beta) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x; // row index of output matrix
+  int32_cuda j = blockIdx.y; // column index of output matrix
+  int32_cuda tgt_index = i * stride; // actual index of data
+  int32_cuda srcA_index = -1, srcB_index = -1, sparse_index = -1;
+  Real tgt_value = beta * data[tgt_index];
+  for (int k = 0; k < srcB_stride; k++) {
+    srcB_index = j * srcB_stride + k;
+    sparse_index = srcB_data[srcB_index].column;
+    if (sparse_index == -1)
```
Comment (danpovey) [inline src/cudamatrix/cu-kernels.cu]: Conditional statements are bad for pipelining.  It's better if you move the conditional out of the loop over k.

File: src/cudamatrix/cu-kernels.cu [hunk 2]
```diff
@@ -748,6 +748,27 @@ static void _add_mat_mat_elements(Real *data, const Real *srcA_data, const Real
     }
 }
 
+template<typename Real>
+__global__
+static void _add_mat_smat(Real *data, int stride, Real alpha, const Real *srcA_data, const int srcA_stride, bool transA, const RowElement<Real>* srcB_data, const int srcB_stride, Real beta) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x; // row index of output matrix
+  int32_cuda j = blockIdx.y; // column index of output matrix
+  int32_cuda tgt_index = i * stride; // actual index of data
```
Comment (danpovey) [inline src/cudamatrix/cu-kernels.cu]: perhaps you mean = i * stride + j;

File: src/cudamatrix/cu-kernels.cu [hunk 3]
```diff
@@ -748,6 +748,27 @@ static void _add_mat_mat_elements(Real *data, const Real *srcA_data, const Real
...
+  int32_cuda srcA_index = -1, srcB_index = -1, sparse_index = -1;
```
Comment (danpovey) [inline src/cudamatrix/cu-kernels.cu]: I don't recommend to declare things before they are used.

**Comment by naxingyu** [comment]
Constructor from SparseMatrix is committed, but not completed. I don't know if the reinterprete_cast method used in CuSparseMatrix is safe. And it seems impossible to convert CuRowSparseMatrix to SparseMatrix since the total number of columns is unavailable. Or is it not necessary to do that?

**Comment by danpovey** [comment]
You definitely need to store the total number of columns as a class member.
Otherwise you wouldn't be able to implement the NumCols() function.
Dan

**Comment by danpovey** [comment]
Regarding the reinterpret_cast, I guess you mean this:

*#endif*

  {

    *reinterpret_cast*<*const* SparseMatrix<Real>*>(*this*
)->CopyToMat(&(M->Mat()),

                                                                 trans);

  }

}

This is a bit subtle.  You need to make sure that the data-members, and
their order, match the members for the SparseMatrix exactly.  If you look
at CuSparseMatrix, you'll see that they do.  Think carefully about what
happens if there is no GPU, when designing this.


Dan

**Comment by naxingyu** [comment]
So I should define both num_cols_ and elements_per_row_ ?

**Comment by danpovey** [comment]
Yes.  They are separate issues: the num_cols is the number of columns in
the matrix it represents sparsely; the  elements_per_row is the maximum
number of nonzero elements in any given row.
Dan

**Comment by naxingyu** [comment]
Most of the parts are finished. Test code is a todo.

**Comment by danpovey** [comment]
Thanks!
I don't have time to check this right now so I'll wait till later--
hopefully some bugs/problems, if they exist will be found by others.
Dan

**Comment by naxingyu** [comment]
I think a SparseMatrix constructor from Matrix is necessary for unit 
test of CuMatrix::AddMatSmat since I plan to AssertEqual the output of 
AddMatMat and AddMatSmat.
Xingyu

**Comment by danpovey** [comment]
You mean a constructor in class Matrix that takes class SparseMatrix?
Initializing a SparseMatrix from a non-sparse Matrix may not always make
sense.
Use function CopyToMat in class SparseMatrix.
You could also add a function CopyFromMat to class SparseMatrix, and maybe
a constructor accepting a Matrix.
But in order for the matrix stuff to be usable without SparseMatrix, I
prefer not to have a constructor in class Matrix that takes SparseMatrix.
Dan

**Comment by naxingyu** [comment]
Oh I could start from a SparseMatrix and make a Matrix to do test. Thanks.
Xingyu

**Comment by naxingyu** [comment]
Unit test code is committed. Speed test throw cu-allocator error, i'm still working on it.

**Comment by danpovey** [comment]
Perhaps your code has a memory leak?
cuda-memcheck may help you debug that.
Dan

**Comment by naxingyu** [comment]
Yes. Thanks.
Xingyu

**Comment by naxingyu** [comment]
I figured out that I was accessing the CuMatrix data in the wrong way.
I did idx = i * stride + j, but it should be idx = i + j * stride since 
CuMatrix is using a column major representation right?
It was weird how this code passed the unit test. I'm doing a full check.
Xingyu

**Comment by danpovey** [comment]
No, CuMatrix is row-major.
However, CUDA BLAS routines require column-major data, so any time we call
one of them we have to switch the transpose-ness.
Dan

**Comment by danpovey** [comment]
I just had a look at your operator = and it's not right.  SEe if you can
spot the problem.
The code is like this:


+template <typename Real>+CuRowSparseMatrix<Real>& CuRowSparseMatrix<Real>::
operator = (+ const CuRowSparseMatrix<Real> &smat) {+#if HAVE_CUDA == 1+ if
(CuDevice::Instantiate().Enabled()) {+ data_ = smat.data_;+ num_rows_ =
smat.num_rows_;+ num_cols_ = smat.num_cols_;+ elements_per_row_ = smat.
elements_per_row_;+ } else

**Comment by naxingyu** [comment]
Are you referring to the self assignment check?

**Comment by danpovey** [comment]
No, to the fact that you just copy the data_ pointer.  You need to
re-allocate the memory and copy, because otherwise when you hit the
destructor, the data_ pointer will be deallocated twice.

**Comment by naxingyu** [comment]
I thought that won't be a problem since operator= is overloaded in 
CuArray...

**Comment by danpovey** [comment]
Oh, sorry, you're right.  I assumed it was a pointer.

**Comment by naxingyu** [comment]
cuda-memcheck told me that there be invalid read of global memory at 
line 758 and 777. Those are where the CuMatrix data is accessed, not the 
CuRowSpareMatrix data. That's why I thought about the column-major 
problem at first. But after knowing what you told me, that won't be a 
problem. There might be something wrong with the grid and block initializer.

**Comment by naxingyu** [comment]
758 and 777 of cu-kernel.cu. Sorry.

**Comment by danpovey** [comment]
OK, I think I see the problem.
CUDA always launches threads in blocks of a certain size, 32 or something,
and you can get kernels running that are out of the range you specified.
So sometimes you have to do a range-check and then return.
E.g. you should probably be checking that the i and j are within range.
 [your j in _add_mat_smat should probably be OK though, because it only
depends on the blockIdx, not the threadIdx.]
Dan

**Comment by naxingyu** [comment]
That solves the problem. Thanks.
Now I'm thinking of moving the -1 index check out of the for loop.
https://github.com/kaldi-asr/kaldi/pull/318/files#diff-ac2b19fd476b7714f65049829e244de0R762

It seems the only way is to change the type of elements_per_row_ from 
MatrixIndexT to CuArray<MatrixIndexT>, is it?

**Comment by danpovey** [comment]
Yes, you'd probably need to do something like that.  If you were to make
that change it would probably make sense to change other aspects  of the
design too, like the choice to pad with -1's in the first place.  I.e. it
would probably make more sense in that case to store the (begin, end) index
into an array of elements as a pair.
Dan

**Comment by naxingyu** [comment]
(The content looks better on Github md)
I am now using another solution, i.e.
```
MatrixIndexT stride_;  // the maximum number of non-zero elements per row
CuArray<MatrixIndexT> elements_per_row_; // the number of non-zero 
elements per row
```
The result of speed test is below
```
LOG (SelectGpuId():cu-device.cc:103) Manually selected to compute on CPU.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 16, speed was 0.525451 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 32, speed was 0.879961 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 64, speed was 1.53888 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 128, speed was 2.70085 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 256, speed was 3.74421 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 512, speed was 4.34207 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 1024, speed was 4.77322 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 16, speed was 0.534731 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 32, speed was 0.844044 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 64, speed was 1.67874 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 128, speed was 2.1621 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 256, speed was 2.37722 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 512, speed was 2.59976 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 1024, speed was 2.77911 gigaflops.
LOG (IsComputeExclusive():cu-device.cc:246) CUDA setup operating under 
Compute Exclusive Mode.
LOG (FinalizeActiveGpu():cu-device.cc:213) The active GPU is [0]: Tesla 
K20m    free:4705M, used:94M, total:4799M, free/total:0.980312 version 3.5
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 16, speed was 0.114005 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 32, speed was 0.695824 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 64, speed was 3.62374 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 128, speed was 5.55943 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 256, speed was 7.17283 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 512, speed was 8.03841 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<float>, for dim = 1024, speed was 7.81052 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 16, speed was 0.143882 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 32, speed was 0.766739 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 64, speed was 3.50573 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 128, speed was 5.74016 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 256, speed was 7.14008 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 512, speed was 7.85815 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For 
CuMatrix::AddMatSmat<double>, for dim = 1024, speed was 6.19297 gigaflops.
```

**Comment by danpovey** [comment]
OK-- I assume that means the speed is better than before?

**Comment by naxingyu** [comment]
(The content looks better on Github md)
I am now using another solution, i.e.

```
MatrixIndexT stride_; // the maximum number of non-zero elements per row
CuArray<MatrixIndexT> elements_per_row_; // the number of non-zero elements per row
```

The result of speed test is below

```
LOG (SelectGpuId():cu-device.cc:103) Manually selected to compute on CPU.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 16, speed was 0.525451 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 32, speed was 0.879961 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 64, speed was 1.53888 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 128, speed was 2.70085 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 256, speed was 3.74421 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 512, speed was 4.34207 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 1024, speed was 4.77322 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 16, speed was 0.534731 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 32, speed was 0.844044 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 64, speed was 1.67874 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 128, speed was 2.1621 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 256, speed was 2.37722 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 512, speed was 2.59976 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 1024, speed was 2.77911 gigaflops.
LOG (IsComputeExclusive():cu-device.cc:246) CUDA setup operating under Compute Exclusive Mode.
LOG (FinalizeActiveGpu():cu-device.cc:213) The active GPU is [0]: Tesla K20m	free:4705M, used:94M, total:4799M, free/total:0.980312 version 3.5
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 16, speed was 0.114005 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 32, speed was 0.695824 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 64, speed was 3.62374 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 128, speed was 5.55943 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 256, speed was 7.17283 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 512, speed was 8.03841 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<float>, for dim = 1024, speed was 7.81052 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 16, speed was 0.143882 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 32, speed was 0.766739 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 64, speed was 3.50573 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 128, speed was 5.74016 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 256, speed was 7.14008 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 512, speed was 7.85815 gigaflops.
LOG (TestCuMatrixMatSmat():cu-matrix-speed-test.cc:118) For CuMatrix::AddMatSmat<double>, for dim = 1024, speed was 6.19297 gigaflops.
```

**Comment by naxingyu** [comment]
Slightly better.

**Comment by danpovey** [comment]
What percentage better?  I don't want extra complexity if it's just a tiny
difference.

**Comment by naxingyu** [comment]
Depends on sparsity. 13.8% faster on average. Largest acceleration is 
53.27%.

**Comment by danpovey** [comment]
OK, fine.
Create a pull request for any recent changes, if needed.
Dan

**Comment by naxingyu** [comment]
Done. I update this one with clean commit history and create another for 
multitree.

File: src/cudamatrix/cu-sparse-matrix.cc [hunk 4]
```diff
@@ -42,6 +42,218 @@
...
+template <typename OtherReal>
+void CuRowSparseMatrix<Real>::CopyToSmat(SparseMatrix<OtherReal> *smat) const {
+  KALDI_ASSERT(smat != NULL);
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    std::vector<RowElement<Real> > cpu_data;
+    data_.CopyToVec(&cpu_data);
+    std::vector<std::vector<std::pair<MatrixIndexT, Real> > > pairs(num_rows_);
+    for (int32 i = 0; i < cpu_data.size(); ++i) {
+      pairs[i].push_back(std::make_pair(cpu_data[i].column, cpu_data[i].weight));
```
Comment (hhadian) [inline src/cudamatrix/cu-sparse-matrix.cc]: @naxingyu, shouldn't this line be something like pairs[i/stride_] ? Currently index 'i' can get out of bounds of pairs.

**Comment by danpovey** [comment]
Guys, I thought we were going to commit this stuff to the 'multitree' branch?  Xingyu, could you please make this a pull request to the 'multitree' branch?  
Dan

**Comment by hhadian** [comment]
We are committing to multitree. I just needed to add a line note and this was the place github let me do it. Currently all latest changes are merged in multitree.

**Comment by danpovey** [comment]
OK thanks.

**Comment by hhadian** [comment]
Thank you. I'm testing SparseLinearComponent in gpu mode now. There are some minor problems such as the line note I added. I'll push tomorrow.

**Comment by danpovey** [comment]
Great- thanks.
Dan

**Comment by naxingyu** [comment]
This stuff has been committed to the 'multitree' branch couple of days 
ago by Hossein and me (several commits mixed).
Should I close this one?
Xingyu

**Comment by danpovey** [comment]
I'll close it.

**Comment by naxingyu** [comment]
Thanks. I'll do a full check of the CuRowSparseMatrix code in multitree, 
since it was pushed asynchronously with master.

**Pull request closed without merging**